src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala

```diff
@@ -341,6 +325,84 @@ class SamRecordClipper(val mode: ClippingMode, val autoClipAttributes: Boolean)
+        case _  => Unit
+      }
+    }
+  }
+
+  /**
+    * Ensures sufficient masking or hard clipping exists on reads that may have soft-clipping. The read will
+    * only be altered if:
+    *   1. ClippingMode is Hard of SoftWithMask
+    *   2. The read is already clipped at the appropriate end
+    *   3. Soft-clipping exist within the first `length` bases of the appropriate end
+    *
+    * If all of those conditions are met, the soft-clipping will be upgraded to masking or hard clipping
+    * up until length bases are clipped or masked. E.g. if the incoming cigar is `10H10S80M` and `length` is
+    * 15, the resulting cigar will be `15H5S80M`.
+    */
+  protected def upgradeClipping(rec: SamRecord, length: Int, fromStart: Boolean): Unit = if (mode != ClippingMode.Soft) {
+    require(length > 0, "Length must be >= 0")
+    def iter = if (fromStart) rec.cigar.iterator else rec.cigar.reverseIterator
+    val hardClipped = iter.takeWhile(_.operator == Op.H).map(_.length).sum
+    val softClipped = iter.dropWhile(_.operator == Op.H).takeWhile(_.operator == Op.S).map(_.length).sum
+
+    if (hardClipped < length && softClipped > 0) {
+      val lengthToUpgrade = math.min(softClipped, length - hardClipped)
+      var (elems, bases, quals) = {
+        if (fromStart) (rec.cigar.coalesce.elems, rec.bases, rec.quals)
+        else (rec.cigar.coalesce.elems.reverse, rec.bases.reverse, rec.quals.reverse)
+      }
+
+      mode match {
+        case ClippingMode.Soft =>
+          unreachable("Should never reach here when mode == Soft")
+        case ClippingMode.SoftWithMask =>
+          hardMaskStartOfRead(bases, quals, lengthToUpgrade)
+        case ClippingMode.Hard =>
+          bases = bases.drop(lengthToUpgrade)
+          quals = quals.drop(lengthToUpgrade)
+          val newElems = ArrayBuffer[CigarElem]()
+
+          elems match {
+            case CigarElem(Op.H, h) +: CigarElem(Op.S, s) +: remaining =>
+              newElems += CigarElem(Op.H, h+lengthToUpgrade)
+              if (s > lengthToUpgrade) newElems += CigarElem(Op.S, s-lengthToUpgrade)
+              newElems ++= remaining
+            case CigarElem(Op.S, s) +: remaining =>
+              newElems += CigarElem(Op.H, lengthToUpgrade)
+              if (s > lengthToUpgrade) newElems += CigarElem(Op.S, s-lengthToUpgrade)
+              newElems ++= remaining
+            case es =>
+              unreachable(s"Cigar $es doesn't contain soft clipping at the start")
+          }
+
+          elems = newElems
+      }
+
+      if (fromStart) {
+        rec.cigar = Cigar(elems)
+        rec.bases = bases
+        rec.quals = quals
+      }
+      else {
+        rec.cigar = Cigar(elems.reverse)
+        rec.bases = bases.reverse
+        rec.quals = quals.reverse
+      }
+
+      clipExtendedAttributes(rec, lengthToUpgrade, fromStart=fromStart)
+    }
```

Comment:
Oh god, it took a long while to read through this and understand all the book-keeping, good work.